Delete user state when they close the tab. Add an unload event for the demo and a delete_callback on gr.State to let developers control how resources are cleaned up (#7829)
gradio-app/gradio: .changeset/better-tires-shave.md

---
"@gradio/client": minor
"gradio": minor
"gradio_client": minor
---

highlight:

> #### Automatically delete state when a user closes the browser tab

Reviewer: Super nit: maybe we should consider rewording in light of the fact that the deletion is not immediate
```suggestion
#### Automatically delete state after user has disconnected from the webpage
```

Gradio now automatically deletes `gr.State` variables stored in the server's RAM when users close their browser tab.
The deletion will happen 60 minutes after the server detected a disconnect from the user's browser.
If the user connects again in that timeframe, their state will not be deleted.

> Additionally, gradio now runs an `unload` event when users disconnect (this does not have a 60 minute delay).

Reviewer: ```suggestion
Additionally, Gradio now includes a `Blocks.unload()` event, allowing you to run cleanup jobs when users disconnect (this does not have a 60 minute delay).
```
